### PR TITLE
NP-33453-update-doi, implemented

### PR DESCRIPTION
**Pull request by truhacevkir**

### CODE REVIEW

**Review by anetteOlli — APPROVED** [review]

File: s3-import-commons/src/main/java/no/unit/nva/publication/s3imports/DoiLinkUpdateEventEmitter.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        logger.info(fileLocation.toString());
+        return s3Driver.getFile(fileLocation);
+    }
+
+    private EventReference parseInput(InputStream input) {
+        String inputString = IoUtils.streamToString(input);
+        return attempt(() -> EventReference.fromJson(inputString))
+                   .orElseThrow();
+    }
+}
```
Comment (brinxmat) [inline s3-import-commons/src/main/java/no/unit/nva/publication/s3imports/DoiLinkUpdateEventEmitter.java:84]: A few things to note here, 
- I have moved the code around a bit to make it clearer what the process is, trying to group related operations
- I got rid of several uses of regular expressions (always a code smell ☠️) using File::lines and String::trim (the latter being what I suspect you wanted)
- I removed commented code (🌧️)
- I formatted the code (🌧️)
- Take a look and see if you agree if it is more readable :) 

```
public class DoiLinkUpdateEventEmitter implements RequestStreamHandler {

    private static final Logger logger = LoggerFactory.getLogger(DoiLinkUpdateEventEmitter.class);
    private final S3Client s3Client;
    private final ResourceService resourceService;

    @JacocoGenerated
    public DoiLinkUpdateEventEmitter() {
        this(defaultS3Client(), ResourceService.defaultService());
    }

    public DoiLinkUpdateEventEmitter(S3Client s3Client, ResourceService resourceService) {
        this.s3Client = s3Client;
        this.resourceService = resourceService;
    }

    @Override
    public void handleRequest(InputStream input, OutputStream output, Context context) throws IOException {
        fetchIdentifiersFromInputEvent(input)
            .map(this::fetchPublication)
            .forEach(this::updatePublication);
    }

    private Publication fetchPublication(SortableIdentifier identifier) {
        return attempt(() -> resourceService.getPublicationByIdentifier(identifier)).orElseThrow();
    }
    
    private void updatePublication(Publication publication) {
        var update = moveDoiToLinkField(publication);
        resourceService.updatePublication(update);
    }

    private Publication moveDoiToLinkField(Publication publication) {
        publication.getEntityDescription().getReference().setDoi(publication.getDoi());
        publication.setDoi(null);
        return publication;
    }

    private Stream<SortableIdentifier> fetchIdentifiersFromInputEvent(InputStream inputStream) {
        var eventReference = parseInput(inputStream);
        var s3Driver = new S3Driver(s3Client, eventReference.extractBucketName());
        var fileLocation = UriWrapper.fromUri(eventReference.getUri()).toS3bucketPath();
        logger.info(fileLocation.toString());
        return s3Driver.getFile(fileLocation)
                   .lines()
                   .map(String::trim)
                   .map(SortableIdentifier::new);
    }

    private EventReference parseInput(InputStream input) {
        var inputString = IoUtils.streamToString(input);
        return attempt(() -> EventReference.fromJson(inputString)).orElseThrow();
    }
}
```

File: s3-import-commons/src/test/java/no/unit/nva/publication/s3imports/DoiLinkUpdateEventEmitterTest.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        var publications = IntStream.range(0, 10)
+                               .boxed()
+                               .map(i -> persist(randomPublication()))
+                               .collect(Collectors.toList());
+        s3Client.putObject(
+            PutObjectRequest.builder().bucket("nve-doi-to-update-bucket").key(DEFAULT_FILE_NAME).build(),
+            RequestBody.fromString(getPublicationInCSVFormat(publications)));
+        return publications;
+    }
+}
```
Comment (brinxmat) [inline s3-import-commons/src/test/java/no/unit/nva/publication/s3imports/DoiLinkUpdateEventEmitterTest.java:120]: - Again, I moved things around again for readability.
- Not sure what the regex (☠️) in the csv conversion was about, I replaced it with something less involved
- A few vars and formatting things too.
- Look over the code and tell me what you think :)

```
class DoiLinkUpdateEventEmitterTest extends ResourcesLocalTest {

    public static final String DEFAULT_FILE_NAME = "results.csv";
    private static final Context context = mock(Context.class);
    private FakeS3Client s3Client;
    private ResourceService resourceService;
    private DoiLinkUpdateEventEmitter handler;

    @BeforeEach
    public void init() {
        super.init();
        this.s3Client = new FakeS3Client();
        this.resourceService = new ResourceService(client, Clock.systemDefaultZone());
        this.handler = new DoiLinkUpdateEventEmitter(s3Client, resourceService);
    }

    @Test
    void shouldMoveDoiValueToLinkField() throws IOException, NotFoundException {
        var publicationsToModify = createPublications();
        var eventReference = createEventReference(publicationsToModify);

        var inputStream = toInputStream(eventReference);
        handler.handleRequest(inputStream, new ByteArrayOutputStream(), context);
        var updatedPublications = fetchUpdatedPublications(publicationsToModify);

        var expectedPublications = createPublicationWithDoiValueInLinkField(publicationsToModify);

        assertThat(expectedPublications, (Every.everyItem(hasProperty("doi", is(equalTo(null))))));
        assertThat(expectedPublications, containsInAnyOrder(updatedPublications.toArray()));
    }

    private EventReference createEventReference(List<Publication> publications) {
        s3Client.putObject(
            PutObjectRequest.builder().bucket("nve-doi-to-update-bucket").key(DEFAULT_FILE_NAME).build(),
            RequestBody.fromString(getPublicationInCSVFormat(publications)));
        return new EventReference(null, null, URI.create("s3://nve-doi-to-update-bucket"
                                                         + "/results.csv"));
    }

    private static String getPublicationInCSVFormat(List<Publication> publications) {
        return publications.stream()
                   .map(Publication::getIdentifier)
                   .map(SortableIdentifier::toString)
                   .collect(Collectors.joining("\n"));
    }

    private static Publication moveDoiToNewField(Publication publication) {
        publication.getEntityDescription().getReference().setDoi(publication.getDoi());
        publication.setDoi(null);
        return publication;
    }

    private List<Publication> fetchUpdatedPublications(List<Publication> publicationsToModify) {
        var publications = publicationsToModify.stream()
                               .map(Publication::getIdentifier)
                               .map(identifier -> attempt(
                                   () -> resourceService.getPublicationByIdentifier(identifier)).orElseThrow())
                               .collect(Collectors.toList());
        publications.forEach(publication -> publication.setModifiedDate(null));
        return publications;
    }

    private Publication persist(Publication publication) {
        var userInstance = UserInstance.fromPublication(publication);
        return Resource.fromPublication(publication).persistNew(resourceService, userInstance);
    }

    private List<Publication> createPublicationWithDoiValueInLinkField(List<Publication> publicationsToModify) {
        var publications = publicationsToModify.stream()
                               .map(DoiLinkUpdateEventEmitterTest::moveDoiToNewField)
                               .collect(Collectors.toList());
        publications.forEach(i -> i.setModifiedDate(null));
        return publications;
    }

    private InputStream toInputStream(EventReference request) {
        return attempt(() -> s3ImportsMapper.writeValueAsString(request))
                   .map(IoUtils::stringToStream)
                   .orElseThrow();
    }

    private List<Publication> createPublications() {
        return IntStream.range(0, 10)
                   .boxed()
                   .map(i -> persist(randomPublication()))
                   .collect(Collectors.toList());
    }
}
```

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: s3-import-commons/src/main/java/no/unit/nva/publication/s3imports/DoiLinkUpdateEventEmitter.java (hunk 1 above)
Comment (truhacevkir) [reply s3-import-commons/src/main/java/no/unit/nva/publication/s3imports/DoiLinkUpdateEventEmitter.java:84]: Agree that this code is more readable :)

File: s3-import-commons/src/test/java/no/unit/nva/publication/s3imports/DoiLinkUpdateEventEmitterTest.java (hunk 2 above)
Comment (truhacevkir) [reply s3-import-commons/src/test/java/no/unit/nva/publication/s3imports/DoiLinkUpdateEventEmitterTest.java:120]: Regex was for removing all necessary brackets from the string when converting List to String :) 
Test is more readable and you have removed all not unnecessary methods

**Pull request merged**